### PR TITLE
Update Litegraph: Canvas Pointer

### PR DESCRIPTION
**Pull request by webfiltered**
## Litegraph CanvasPointer API

Major update to the way pointer intraction is handled.

<details open>

<summary>Video pointer explanation</summary>

- Circles: Left, middle, and right mouse buttons
- Control, Alt, and Shift modifier keys
- Blue square flash: undo / redo
image

</details>

## Videos

<details open>

<summary>Videos</summary>

#### Click vs Drag

Huge complexity with the middle mouse button, here:
- Short click is click
- Long click is click
- Dragging is dragging

https://github.com/user-attachments/assets/6ed08187-6d70-4134-8bbd-9dd99e77c962

#### Double click & click drift

- Distance between clicks matters
- Where the clicks happen matters
- Click drift buffer is noticeable here - the node doesn't move, despite fast (and imperfect) movement + clicking

https://github.com/user-attachments/assets/f519118c-8077-41ea-83b9-c773697e2158

#### Area select & deselect

- Control + drag to area select, replacing the current selection
- After starting an area selection, you can
  - Hold `Shift` to add items to the selection
  - Hold `Alt` to remove items from the selection

https://github.com/user-attachments/assets/c85a2a35-d6fd-4e59-8d92-5c2789e5220b

- Easier to use shortcuts are planned

</details>

### Frontend code changes
- Adds settings options for pointer configuration
  - The two marked experimental would be great candidates for an "advanced" section - or in an expander, hidden by default
- Update touch device API - rename
- Fixes exception thrown on double-clicking slot
- Requires https://github.com/Comfy-Org/litegraph.js/pull/308
- Requires https://github.com/Comfy-Org/litegraph.js/pull/315
- Best served with #1572

### CODE REVIEW

**Review by huchenlei — CHANGES_REQUESTED** [review]
https://github.com/Comfy-Org/ComfyUI_frontend/blob/6fefcaad7bee39d2928a6a808ec62226701b34ba/src/extensions/core/snapToGrid.ts#L85-L90

is using `node_dragged` removed in litegraph. Can you adjust it to use the new pointer API? I think eventually we should have this feature upstreamed to litegraph.

**Comment by webfiltered** [comment]
Probably more than you wanted.  But there were "a few small problems" just trying to quickly patch snap to grid.  So I rewrote the whole thing.

https://github.com/Comfy-Org/ComfyUI_frontend/pull/1572
https://github.com/Comfy-Org/litegraph.js/pull/315

Let me know if you need me to add little interim PRs - to make pass at least the compile tests without simultaneously merging both repos.

File: src/stores/coreSettings.ts [hunk 1]
```diff
@@ -528,5 +528,49 @@ export const CORE_SETTINGS: SettingParams[] = [
     type: 'boolean',
     defaultValue: true,
     versionAdded: '1.4.0'
+  },
+  {
+    id: 'Comfy.Pointer.ClickDrift',
```
Comment (huchenlei) [inline src/stores/coreSettings.ts:540]: I am going to add a new top level category for all litegraph related features. The Comfy category now is very jammed.

**Review by huchenlei — COMMENTED** [review]
Do you have any idea why those jest test fail? It seems like in jest test environment `Element` don't have `hasPointerCapture` method.

For playwright tests, all screenshot diffs are due to node/group selection state change, which are expected, other than the changeTracker tests. I verified locally that bypass and undo bypass works fine, but it is just not getting through the tests. We can disable these flaky undo/redo tests for now.
image

File: src/stores/coreSettings.ts (hunk 1 above)
Comment (webfiltered) [reply src/stores/coreSettings.ts:540]: For sure!  I was thinking the settings page needs tabs or expanders.

**Comment by webfiltered** [comment]
> It seems like in jest test environment `Element` don't have `hasPointerCapture` method.

Yup!  That's why I wanted to try vitest - you pick your own jsdom version, and apparently the latest one supports it.  When using vitest locally I am still failing a few tests - passing most in `fast`.

**Comment by webfiltered** [comment]
Manually reviewed all failing tests.  Most just needed click-related updates because of the changes.

The only difference that I'm not 100% on is widgets; maybe clicking any widget should select the node.  Community testing as-is would be good, though.

@huchenlei If you're happy, I believe it's ready for expectation updates.

**Review by huchenlei — APPROVED** [review]

**Comment by yoland68** [comment]
This PR is causing rendering litegraph rendering issue for me:
Screenshot 2024-11-18 at 8 36 02 AM
Currently:
Screenshot 2024-11-18 at 8 37 14 AM


Before: (2a93f87)


Screenshot 2024-11-18 at 8 37 58 AM

**Comment by rgthree** [comment]
Note, this pull does break widgets' `mouse` event handlers no longer receiving `pointerup` events, which seems unintentional.

I think I have an (untested) solution in https://github.com/Comfy-Org/litegraph.js/issues/341#issuecomment-2496209397

**Pull request merged**